src/language-markdown/print-whitespace.js

```diff
@@ -0,0 +1,282 @@
+ *
+ * @param {AstPath} path current position in nodes tree
+ * @returns {boolean} `true` if Space tends to be inserted between these types of letters, `false` otherwise.
+ */
+function isInSentenceWithCJSpaces(path) {
+  const sentenceNode = path.parent;
+  if (sentenceNode.usesCJSpaces !== undefined) {
+    return sentenceNode.usesCJSpaces;
+  }
+
+  const stats = { " ": 0, "": 0 };
+
+  for (let i = 1; i < sentenceNode.children.length - 1; ++i) {
+    const node = sentenceNode.children[i];
+    if (
+      node.type === "whitespace" &&
+      (node.value === " " || node.value === "")
+    ) {
+      const previousKind = sentenceNode.children[i - 1].kind;
+      const nextKind = sentenceNode.children[i + 1].kind;
+      if (
+        (previousKind === KIND_CJ_LETTER && nextKind === KIND_NON_CJK) ||
+        (previousKind === KIND_NON_CJK && nextKind === KIND_CJ_LETTER)
+      ) {
+        ++stats[node.value];
+      }
+    }
+  }
+
+  // Injects a property to cache the result.
+  sentenceNode.usesCJSpaces = stats[" "] > stats[""];
+  return sentenceNode.usesCJSpaces;
+}
+
+/**
+ * @typedef {import("./utils.js").WordNode} WordNode
+ * @typedef {import("./utils.js").WhitespaceValue} WhitespaceValue
+ * @typedef {{ next?: WordNode | null, previous?: WordNode | null }} AdjacentNodes
+ * Adjacent node to `WhitespaceNode`. the consecution of `WhitespaceNode` is a bug, so adjacent nodes must be `WordNode`.
+ * @typedef {import("./utils.js").WordKind} WordKind
+ * @typedef {import("../common/ast-path.js").default} AstPath
+ */
+
+/**
+ * Checks if given `"\n"` node can be converted to Space
+ *
+ * For example, if you would like to squash the multi-line string `"You might want\nto use Prettier."` into a single line,
+ * you would replace `"\n"` with `" "`. (`"You might want to use Prettier."`)
+ *
+ * However, you should note that Chinese and Japanese do not use U+0020 Space to divide words, so U+000A End of Line must not be replaced with it.
+ * Behavior in other languages (e.g. Thai) will not be changed because there are too much things to consider. (PR welcome)
+ *
+ * @param {AstPath} path path of given node
+ * @param {AdjacentNodes | undefined} adjacentNodes adjacent sibling nodes of given node
+ * @returns {boolean} `true` if given node can be convertedToSpace, `false` if not (i.e. newline or empty character)
+ */
+function lineBreakCanBeConvertedToSpace(path, adjacentNodes) {
+  if (
+    // no adjacent nodes - this happens for linkReference and imageReference nodes
+    !adjacentNodes ||
+    // e.g. " \nletter"
+    !adjacentNodes.previous ||
+    !adjacentNodes.next
+  ) {
+    return true;
+  }
+
+  const previousKind = adjacentNodes.previous.kind;
+  const nextKind = adjacentNodes.next.kind;
+
+  // "\n" between not western or Korean (han, kana, CJK punctuations) characters always can be converted to Space
+  // Korean hangul simulates latin words; see #6516 (https://github.com/prettier/prettier/issues/6516)
```

Comment:
Why does this comment mention "han, kana, CJK punctuations" if it checks only "isWesternOrKorean"?

Reply:
It's saying the opposite... Thank you for finding it.